### ISSUE TITLE
Tables that subtype AbstractMatrix are not fully supported

### ISSUE THREAD

**Issue by LilithHafner** [JuliaData/Tables.jl#316]
I have a type that subtypes AbstractMatrix _and_ conforms to the Tables.table interface (I think). Yet `Tables.rows(::MyType)` errors, dispatching to https://github.com/JuliaData/Tables.jl/blob/db5e0903cf40a0803ec23ebd92753b70feeab169/src/matrix.jl#L3 without ever checking if my table conforms to the Tables.table interface.

<details>
<summary>Here's a large MWE, I can provide a smaller one upon request</summary>

```julia
using AxisArrays, Tables

abstract type AbstractMatrixTable{T} <: AbstractMatrix{T} end

struct MatrixTable{T, U <: AxisArray{T, 2}} <: AbstractMatrixTable{T}
    data::U
    MatrixTable(data::AxisArray{T, 2}) where T = new{T, typeof(data)}(data)
end
function MatrixTable(data::AbstractMatrix, rows::AbstractVector, cols::Vector{Symbol}; check_unique::Bool=true)
    check_unique && !allunique(cols) && throw(ArgumentError("Column names must be unique"))
    MatrixTable(AxisArray(data; rows, cols))
end

# getter method to avoid conflict with custom getproperty method
data(m::MatrixTable) = getfield(m, :data)

# named axes
named_axes(m::AbstractMatrixTable) = getproperty.(data(m).axes, :val)
rows(m::AbstractMatrixTable) = named_axes(m)[1]
cols(m::AbstractMatrixTable) = named_axes(m)[2]

# Access by dot syntax `data.var3`
Base.propertynames(m::AbstractMatrixTable) = cols(m)
Base.getproperty(m::AbstractMatrixTable, col::Symbol) = getindex(m, :, col)

# Conform to the AbstractMatrix interface
# https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array

# Required
Base.size(m::AbstractMatrixTable) = size(data(m))
function Base.getindex(m::AbstractMatrixTable, inds...) # Access keeps the same type (uncommon design decision); with and exception for m[today(),:b] === 7
    a = data(m)
    ix = AxisArrays.to_index(a, inds...)
    all(i -> i isa Integer, ix) && return AxisArrays.getindex_converted(a, ix...)
    MatrixTable(AxisArrays.getindex_converted(a, (i isa Integer ? (i:i) : i for i in ix)...))
end
Base.setindex!(m::AbstractMatrixTable, val, inds...) = setindex!(data(m), val, inds...)

# Optional
Base.similar(m::AbstractMatrixTable, T::Type=eltype(m)) = MatrixTable(similar(data(m), T), rows(m), propertynames(m), check_unique=false)

# Conform to the Tables.jl interface
# https://tables.juliadata.org/stable/#Implementing-the-Interface-(i.e.-becoming-a-Tables.jl-source)

# Required as a Table
Tables.istable(::Type{<:AbstractMatrixTable}) = true
Tables.columnaccess(::Type{<:AbstractMatrixTable}) = true
Tables.columns(m::AbstractMatrixTable) = m

# Required as a Tables.AbstractColumns
Tables.getcolumn(m::AbstractMatrixTable, i::Int) = data(m)[:, i]
Tables.getcolumn(m::AbstractMatrixTable, s::Symbol) = data(m)[:, s]

# Tests
m = MatrixTable(rand(5,3), 1:5, [:a, :b, :c])
Tables.rows(m)
#=
ERROR: ArgumentError: a 'MatrixTable{Float64, AxisArray{Float64, 2, Matrix{Float64}, Tuple{Axis{:rows, UnitRange{Int64}}, Axis{:cols, Vector{Symbol}}}}}' is not a table; see `?Tables.table` for ways to treat an AbstractVecOrMat as a table
Stacktrace:
 [1] rows(m::MatrixTable{Float64, AxisArray{Float64, 2, Matrix{Float64}, Tuple{Axis{:rows, UnitRange{Int64}}, Axis{:cols, Vector{Symbol}}}}})
   @ Tables ~/.julia/packages/Tables/T7rHm/src/matrix.jl:3
 [2] top-level scope
   @ REPL[263]:1
=#
```

</details>

**Comment by quinnj** [comment]
Sorry for the slow response here; yeah, things are a tad awkward for `AbstractMatrix` subtypes because we're trying to help people who accidentally pass a plain `Matrix` to table sinks. One solution is to just define:

```julia
Tables.rows(x::MyAbstractMatrix) = Tables.RowIterator(Tables.columns(x), Tables.rowcount(Tables.columns(x)))
```

**Comment by quinnj** [comment]
Actually, I think we can support this pretty easily; here's a PR: https://github.com/JuliaData/Tables.jl/pull/319

**Comment by LilithHafner** [comment]
Fixed by #319
